### PR TITLE
Database event store

### PR DESCRIPTION
**Pull request by CodeReaper**

### CODE REVIEW

File: libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore/EventStore/Database/DatabaseEventStore.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace EnergyOriginEventStore.EventStore.Database;
+
+public class DatabaseEventStore : IEventStore
+{
+    private readonly DatabaseEventContext context;
+
+    public DatabaseEventStore(DatabaseEventContext context) => this.context = context;
+
+    #region IEventStore
```
Comment (MartinSchmidt) [inline libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore/EventStore/Database/DatabaseEventStore.cs:12]: fluff

File: libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore/EventStore/Database/DatabaseEventConsumerBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace EnergyOriginEventStore.EventStore.Database;
+
+internal class DatabaseEventConsumerBuilder : IEventConsumerBuilder
+{
+    private readonly Dictionary<Type, IEnumerable<Action<Event<EventModel>>>> handlers = new();
+    private Action<string, Exception>? exceptionHandler;
+    private readonly DatabaseEventContext context;
+    private readonly string topicPrefix;
+    private string? pointer;
```
Comment (MartinSchmidt) [inline libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore/EventStore/Database/DatabaseEventConsumerBuilder.cs:12]:
```suggestion
    private string pointer = "0";
```

File: libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore/EventStore/Database/DatabaseEventConsumer.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    public void Dispose()
+    {
+        timer.Dispose();
+        GC.SuppressFinalize(this);
+    }
+
+    private async Task Work()
+    {
+        await semaphore.WaitAsync();
```
Comment (MartinSchmidt) [inline libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore/EventStore/Database/DatabaseEventConsumer.cs:45]: remove semaphore!
move do while loop into Work() -> rename to worker

File: libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore/EventStore/Database/DatabaseEventConsumer.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+        timer = new PeriodicTimer(TimeSpan.FromMilliseconds(100));
+
+        Task.Run(async () =>
+        {
+            do
+            {
+                await Work();
+            } while (await timer.WaitForNextTickAsync());
+        });
```
Comment (MartinSchmidt) [inline libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore/EventStore/Database/DatabaseEventConsumer.cs:34]:
```suggestion
        Task.Run(async worker());
```

File: libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore.Tests/SqliteDatabaseEventContext.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    async Task DatabaseEventContext.Add(Message message) => await connection.ExecuteAsync("INSERT INTO messages VALUES(NULL, @topic, @payload);", new { topic = message.Topic, payload = message.Payload });
+
+    async Task<Message?> DatabaseEventContext.NextAfter(long? id, string? topicPrefix) => (id, topicPrefix) switch
+    {
+        (null, null) => await connection.QueryFirstOrDefaultAsync<Message?>("SELECT id, topic, payload FROM messages ORDER BY id ASC LIMIT 1;"),
+        (_, null) => await connection.QueryFirstOrDefaultAsync<Message?>("SELECT id, topic, payload FROM messages WHERE id > @id ORDER BY id ASC LIMIT 1;", new { id }),
+        (null, _) => await connection.QueryFirstOrDefaultAsync<Message?>("SELECT id, topic, payload FROM messages WHERE topic LIKE @prefix ORDER BY id ASC LIMIT 1;", new { prefix = $"{topicPrefix}%" }),
+        (_, _) => await connection.QueryFirstOrDefaultAsync<Message?>("SELECT id, topic, payload FROM messages WHERE id > @id AND topic LIKE @prefix ORDER BY id ASC LIMIT 1;", new { id, prefix = $"{topicPrefix}%" })
+    };
```
Comment (MartinSchmidt) [inline libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore.Tests/SqliteDatabaseEventContext.cs:28]: cleanup

File: libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore.Tests/EventStoreTests.cs [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+[Collection("Database collection")]
+public class EventStoreTests : IClassFixture<EventStoreTests.DatabaseFixture>, IDisposable
+{
+    #region Setup
+
+    private readonly DatabaseFixture fixture;
+
+    public EventStoreTests(DatabaseFixture fixture) => this.fixture = fixture;
+
+    private static IEnumerable<object[]> Data() => new List<object[]>
```
Comment (MartinSchmidt) [inline libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore.Tests/EventStoreTests.cs:19]: Find better name

File: libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore.Tests/EventStoreTests.cs [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+        };
+
+    #endregion
+
+    #region Tests
+
+#pragma warning disable IDE0060, xUnit1026
+
+    [Theory]
+    [MemberData(nameof(Data))]
```
Comment (MartinSchmidt) [inline libraries/dotnet/EnergyOriginEventStore/EnergyOriginEventStore.Tests/EventStoreTests.cs:33]: try with inlinedata...

**Review by MartinSchmidt — CHANGES_REQUESTED** [review]

**Comment by CodeReaper** [comment]
Notice I removing FlatFile as a viable option for event store. Only memory event store and database event store is considered stable.

**Review by Exobitt — DISMISSED** [review]

**Review by MartinSchmidt — DISMISSED** [review]

**Review by Exobitt — APPROVED** [review]

**Pull request merged**